tutorials/hyperspectral/Plot-Hyperspectral-Spectra.Rmd

```diff
@@ -131,11 +120,6 @@ aPixeldf$Wavelength <- wavelengths
 
 head(aPixeldf)
 
```

Comment:
Curiosity, not a need to change – why has the H5close() been removed from several of these tutorials?

Reply:
I looked into this behavior closely while I was testing these tutorials, and I believe that there were several instances of that function that were unnecessary. My overall impression is that this is more important when writing to H5 files because it opens a channel for editing. All of the functions in this tutorial are read-only. If you use the `getAnywhere()` function to look at the `h5read()` function, you can see that it actually does call `h5close()` at the end of the function as shown below:

> getAnywhere(h5read)
A single object matching ‘h5read’ was found
It was found in the following places
  package:rhdf5
  namespace:rhdf5
with value

function (file, name, index = NULL, start = NULL, stride = NULL, 
    block = NULL, count = NULL, compoundAsDataFrame = TRUE, callGeneric = TRUE, 
    read.attributes = FALSE, drop = FALSE, ..., native = FALSE) 
{
    loc = h5checktypeOrOpenLoc(file, readonly = TRUE, native = native)
    on.exit(h5closeitLoc(loc))
    if (!H5Lexists(loc$H5Identifier, name)) {
        stop("Object '", name, "' does not exist in this HDF5 file.")
    }
    else {
        oid = H5Oopen(loc$H5Identifier, name)
        type = H5Iget_type(oid)
        num_attrs = H5Oget_num_attrs(oid)
        if (is.na(num_attrs)) {
            num_attrs = 0
        }
        **H5Oclose(oid)**
        if (type == "H5I_GROUP") {
            gid <- H5Gopen(loc$H5Identifier, name)
            obj = h5dump(gid, start = start, stride = stride, 
                block = block, count = count, compoundAsDataFrame = compoundAsDataFrame, 
                callGeneric = callGeneric, ...)
            **H5Gclose(gid)**
        }
        else if (type == "H5I_DATASET") {
            h5dataset <- H5Dopen(loc$H5Identifier, name)
            obj <- h5readDataset(h5dataset, index = index, start = start, 
                stride = stride, block = block, count = count, 
                compoundAsDataFrame = compoundAsDataFrame, drop = drop, 
                ...)
            if (storage.mode(obj) == "character" && H5Aexists(h5dataset, 
                name = "as.na")) {
                if (any(obj == "NA")) {
                  obj[obj == "NA"] <- NA_character_
                }
            }
            try({
                **H5Dclose(h5dataset)**
            })
            cl <- attr(obj, "class")
            if (!is.null(cl) & callGeneric) {
                if (exists(paste("h5read", cl, sep = "."), mode = "function")) {
                  obj <- do.call(paste("h5read", cl, sep = "."), 
                    args = list(obj = obj))
                }
            }
        }
        else {
            message("Reading of object type not supported.")
            obj <- NULL
        }
        if (read.attributes & (num_attrs > 0) & !is.null(obj)) {
            for (i in seq_len(num_attrs)) {
                A = H5Aopen_by_idx(loc$H5Identifier, n = i - 
                  1, objname = name)
                attrname <- H5Aget_name(A)
                if (attrname != "dim") {
                  attr(obj, attrname) = H5Aread(A)
                }
                **H5Aclose(A)**
            }
        }
    }
    obj
}
<bytecode: 0x7ff523385128>
<environment: namespace:rhdf5>
>